ibis/expr/operations.py

```diff
@@ -1608,50 +1550,42 @@ def __init__(self, *args, **kwargs):
 
 
 class AsOfJoin(Join):
-    _arg_names = ['left', 'right', 'predicates', 'by']
-
-    def __init__(self, left, right, predicates, by_predicates):
-        _validate_join_tables(left, right)
-        (self.left,
-         self.right,
-         self.predicates) = _make_distinct_join_predicates(
-            left, right, predicates)
-        self.by_predicates = _clean_join_predicates(
-            self.left, self.right, by_predicates)
+    left = Arg(rlz.noop)
+    right = Arg(rlz.noop)
+    predicates = Arg(rlz.noop)
+    by = Arg(rlz.noop, default=None)
 
-        Node.__init__(
-            self, [self.left, self.right, self.predicates, self.by_predicates])
+    def __init__(self, left, right, predicates, by):
+        super(AsOfJoin, self).__init__(left, right, predicates)
+        self.by = _clean_join_predicates(self.left, self.right, by)
 
 
 class Union(TableNode, HasSchema):
+    left = Arg(rlz.noop)
```

Comment:
Should probably be `TableExpr` here, but leave it as is if making that change doesn't trivially work.